RPIPs/RPIP-47.md

```diff
@@ -26,34 +26,30 @@ Additionally, the pDAO must have the ability to effect change in ways that benef
 
 ## Specification
-### Megapool delegates
-- `Megapool delegate` contracts SHALL have an expiration block (ie, execution layer block)
-- When a new `megapool delegate` is released, the contract upgrade SHALL include:
-  - Setting the new `megapool delegate`'s expiration block to "no expiration"
-  - Setting the previous `megapool delegate`'s expiration block to occur `delegate_upgrade_buffer` after the upgrade
-- A node operator SHALL be able to upgrade their active `megapool delegate` to the latest `megapool delegate` at any time (including after the active `megapool delegate`'s expiration block)
-- After the active `megapool delegate`'s expiration block:
-  - Functions in the active megapool delegate SHALL revert when not called by a protocol contract
-  - No capital SHALL be distributed to the withdrawal address(es)
-- The initial setting of `delegate_upgrade_buffer` SHALL be 4 months.
-
-### Megapool recovery delegate
-- There SHALL be a single `megapool recovery delegate` contract that is used by all megapools in the protocol
-- The `megapool recovery delegate` for a megapool SHALL NOT be usable by anyone if the current block is less than or equal to the active `megapool delegate`'s expiration block
-- The `megapool recovery delegate` for a megapool SHALL be usable by any address if the current block is greater than the active `megapool delegate`'s expiration block
-- The `megapool recovery delegate` SHALL have the following capabilities:
-  - Force exiting ALL validators under the megapool
-    - In the Saturn 1 release, this MAY be a non-functional stub
-  - Removal of any pending validators from the Rocket Pool queue
-  - Processing of voluntary exit proofs
-  - Collection of any penalties or deficits
-  - Final distribution of rewards and capital 
-- The `megapool recovery delegate` SHOULD be updated only rarely, and only to support changes to the Ethereum execution layer exit process, or major changes in the Rocket Pool protocol. Simultaneous upgrades to both the `megapool recovery delegate` and the latest `megapool delegate` SHOULD be avoided if possible.
+
+### Upgrade Buffer Enforcement
+- Each `Megapool delegate` contract SHALL have an execution layer `expiration_block` variable which should initialize to "no expiration"
+- Each `Megapool delegate` contract SHALL have a non-modifiable `upgrade_buffer` value
+  - The `Megapool delegate`'s `upgrade_buffer` SHALL be 120 days
+- The only permitted modification to a `Megapool delegate`'s `expiration_block` variable SHALL be for any protocol contract to set its value to now + its `upgrade_buffer`
+
+### Upgrade
+- When a new `megapool delegate` is released, a protocol contract SHOULD update all previous `megapool delegate`s `expiration_block`'s to now + their `upgrade_buffer`s
```

Comment:
Like, the whole point of enforcing it here is that its enforcing it _against_ changes in the RP governance process. 

In usual practice, I can't see that this is even going to have much of an impact. I strongly doubt we're ever going to be releasing new and different delegates within the 120 day limit. 

So in usual circumstances this has zero effect, as there is only ever likely to be a single active delegate version. 
However, it also gives NO's some protection against unusual circumstances.

Reply:
I think you're seeing this differently than I. It's not a grace period at all to me. It's an observation period to see the delegate operating in the real world and getting lindy. 

So the situation is:
- I'm on delegate X on day 0
- Delegate Y is released
- I begin observing, with up to 120 days available
- After 60 days, delegate Z is proposed

My options are now:
- Decide I'm comfortable enough to go with delegate Y before Z is live
  - This would give me until day 180 to make a decision on Z
- Let Z go live; before day 120:
  - Decide I'm comfortable with Z
  - Decide I'm not comfortable and exit

This all seems ok to me. In edge cases a Node Operator may feel the need to exit, but they aren't forced into anything ever.

----
Your variant is reasonably similar:
- I'm on delegate X on day 0
- Delegate Y is released
- I begin observing, with up to 120 days available
- After 60 days, delegate Z is proposed

My options are now:
- Decide I'm comfortable enough to go with delegate Y before Z is live
  - This would give me until 180 to make a decision on Z
- Let Z go live; before day **180** <--- this is the only change:
  - Decide I'm comfortable with Z
  - Decide I'm not comfortable and exit
----

Reply:
Hmm. I agree there is not much difference in the usual case. But I do feel the 'all active' version is preferable for the unusual case with little downside.

Consider the following, as a small alteration to your first sccenario:

I'm on delegate X on day 0
* Delegate Y is released
* I begin observing, with up to 120 days available
* After 119 days, protocol is upgraded to delegate Z (I know this is unlikely in practice)

My options are:
* Decide I don't like Y, and bail with plenty of time (all good here)
* Decide I'm comfortable enough to upgrade to delegate Y before Z is live, (all good here too).
* Decide I'm comfortable with delegate Y, but I'm lazy, so I don't manually upgrade. When Z comes in on day 119, I have effectively no time to evaluate, and I'm moved to Z, possibly before I even notice Z has been proposed. **This is very bad.**

---

It just seems very low cost to me to prevent the third scenario by resetting the grace/observation period. Yes it's unlikely to occur, but resetting active delegate versions (when it's also very unlikely they'll be more than one at a time) really seems like an easy way to just let NO's never need to worry about this.

Reply:
I'll update the wording to some variation of 'delegate versions currently in use', tomorrow, but I do feel that's a better version than just updating the last delegate.

Reply:
Hey, yes we agree with you @LongForWisdom. Some variation of 'delegate versions currently in use' or 'active' would be fine. We were a little worried about implementation but have review and are confident it can be achieved.